python/mlx/optimizers.py

```diff
@@ -498,3 +497,155 @@ def apply_single(
+          denominator to improve numerical stability. Default: ``1e-8``
+
+    Args:
+        learning_rate (Optional[float], optional): The learning rate
+            :math:`\alpha`. Defaults to None.
+        eps (Tuple[float, float], optional): The first term :math:`\epsilon`
+            added to the square of the gradients to improve numerical
+            stability and the second term :math:`\epsilon` is used for
+            parameter scaling if `parameter_scale` is set to `True`.
+            Defaults to (1e-30, 1e-3).
+        clip_threshold (float, optional): Clips the unscaled update at
+            `clip_threshold`. Defaults to 1.0.
+        decay_rate (float, optional): Controls the rate of increasing
+            :math:`\beta_2`. Defaults to -0.8.
+        beta_1 (Optional[float], optional): If set to a value bigger than zero
+            then first moment will be used. Defaults to None.
+        weight_decay (float, optional): The weight decay :math:`\lambda`.
+            Defaults to 0.0.
+        scale_parameter (bool, optional): If set to `True` the learning rate
+            will be scaled by :math:`max(\epsilon_1, RMS(X_{t-1}))`.
+            Defaults to True.
+        relative_step (bool, optional): If set to `True` the `learning_rate`
+            will be ignored and relative step size will be computed.
+            Defaults to True.
+        warmup_init (bool, optional): If set to True then relative step size
+            will be calculated by the current step. Defaults to False.
+    """
+
+    def __init__(
+        self,
+        learning_rate: Optional[float] = None,
+        eps: Tuple[float, float] = (1e-30, 1e-3),
+        clip_threshold: float = 1.0,
+        decay_rate: float = -0.8,
+        beta_1: Optional[float] = None,
+        weight_decay: float = 0.0,
+        scale_parameter: bool = True,
+        relative_step: bool = True,
+        warmup_init: bool = False,
+    ):
+        super().__init__()
+        self.learning_rate = learning_rate
+        self.eps = eps
+        self.clip_threshold = clip_threshold
+        self.decay_rate = decay_rate
+        self.beta_1 = beta_1
+        self.weight_decay = weight_decay
+        self.scale_parameter = scale_parameter
+        self.relative_step = relative_step
+        self.warmup_init = warmup_init
+
+    def compute_rms(self, inputs):
+        return mx.sqrt(mx.mean(mx.square(inputs)))
+
+    def compute_learning_rate(self, step, rms):
+        relative_step_size = self.learning_rate
+        if self.relative_step:
+            min_step = 1e-6 * step if self.warmup_init else 1e-2
+            relative_step_size = mx.minimum(min_step, 1 / mx.sqrt(mx.array(step)))
+
+        parameter_scale = 1.0
+        if self.scale_parameter:
+            parameter_scale = mx.maximum(mx.array(self.eps[1]), rms)
+        return parameter_scale * relative_step_size
+
+    def approximate_exp_moving_avg(self, exp_avg_sq_row, exp_avg_sq_col):
```

Comment:
These should probably be `_function_name` since they are implementation details.